### DISCUSSION TITLE
Dreaded "Unable to load DLL libSkiaSharp" error

### DISCUSSION THREAD

**Discussion by bigpilot** [QuestPDF/QuestPDF#719] [General]
I've installed the QuestPDF NuGet package and I'm getting this error. The SkiaSharp.dll is in the output directory but there's no "libSkiaSharp" DLL. I assume these are one and the same?

Here are the packages I've installed:

```
<?xml version="1.0" encoding="utf-8"?>
<packages>
  <package id="EntityFramework" version="6.4.4" targetFramework="net461" />
  <package id="HarfBuzzSharp" version="7.3.0" targetFramework="net461" requireReinstallation="true" />
  <package id="HarfBuzzSharp.NativeAssets.macOS" version="7.3.0" targetFramework="net461" requireReinstallation="true" />
  <package id="HarfBuzzSharp.NativeAssets.Win32" version="7.3.0" targetFramework="net461" requireReinstallation="true" />
  <package id="QuestPDF" version="2023.10.2" targetFramework="net462" />
  <package id="SkiaSharp" version="2.88.6" targetFramework="net461" requireReinstallation="true" />
  <package id="SkiaSharp.HarfBuzz" version="2.88.6" targetFramework="net461" requireReinstallation="true" />
  <package id="SkiaSharp.NativeAssets.macOS" version="2.88.6" targetFramework="net461" requireReinstallation="true" />
  <package id="SkiaSharp.NativeAssets.Win32" version="2.88.6" targetFramework="net461" requireReinstallation="true" />
  <package id="System.Buffers" version="4.5.1" targetFramework="net461" />
  <package id="System.Memory" version="4.5.5" targetFramework="net461" />
  <package id="System.Numerics.Vectors" version="4.5.0" targetFramework="net461" />
  <package id="System.Runtime.CompilerServices.Unsafe" version="6.0.0" targetFramework="net461" />
</packages>
```
I noticed there are no References to the native assets in my application. 

Why isn't this working when I install the required NuGet packages?

**Comment by MarcinZiabek** [comment]
It is indeed dreaded! 🤣 Hopefully [this effort](https://github.com/QuestPDF/QuestPDF/discussions/622) with give QuestPDF more control over native dependencies. Though, I am sure that all these problems are not a coincidence and loading native code must just be difficult cross-platform.

Please share more information about your OS and runtime.

Have you installed these nugets in your startup project? (https://github.com/QuestPDF/QuestPDF/issues/694)

**Reply by bigpilot** [reply]
I personally don't see how a custom SkiaSharp version (or own Skia wrapper) would solve these DLL problems. 

The problems I'm having seem more like a packaging or configuration error.

**Reply by MarcinZiabek** [reply]
There are additional benefits when having a custom-built Skia. We will see how well the effort goes 😀

**Reply by MarcinZiabek** [reply]
> Please share more information about your OS and runtime.
> Have you installed these nugets in your startup project? (https://github.com/QuestPDF/QuestPDF/issues/694)

Please provide those details. What type of project are you building?

**Reply by bigpilot** [reply]
I'm using Windows 10 and .NET 4.6.2 as runtime.

My application type is WinForms and I'm using QuestPDF as a report generator.

**Comment by bigpilot** [comment]
libSkiaSharp.dll and libHarfBuzzSharp.dll aren't in my Debug output directory but they are present in my Packages directory.

\packages\SkiaSharp.NativeAssets.Win32.2.88.6\runtimes\win-x64\native\libSkiaSharp.dll
\packages\HarfBuzzSharp.NativeAssets.Win32.7.3.0\runtimes\win-x64\native\libHarfBuzzSharp.dll

**Reply by MarcinZiabek** [reply]
What CPU target are you using? `Any` / `x86` / `x64`? I would try with `x64`.

**Reply by bigpilot** [reply]
I tried building with x64 but same result.

**Comment by bigpilot** [comment]
I'm trying to build QuestPDF and SkiaSharp myself. 

Building QuestPDF was no problem, but only results in a QuestPDF.dll compilation artifact.

Building SkiaSharp by running the build.ps1 script gave me this output:

> ========================================
> git-sync-deps
> ========================================
> An error occurred when executing task 'git-sync-deps'.
> Error: The libskia C++ API version did not match the expected '115', instead was ''.
> An error occurred when executing task 'externals-android'.
> Error: .NET CLI: Process returned an error (exit code 1).

**Reply by MarcinZiabek** [reply]
Turns out that building Skia is quite a terrible experience for everybody, especially for a person used to dotnet workflows and IDEs 🤣 

I have made some progress in building the library using GitHub Actions, please take a look: https://github.com/MarcinZiabek/skia-custom/blob/main/.github/workflows/build.yml It still fails on Windows and I need to analyze more which flag could cause it. It is pure pain and frustration but I start to understand what's going on.

**Comment by bigpilot** [comment]
Sorry but this is taking me too long to get working. I've switched to PDFSharp and [MigraDoc](http://www.pdfsharp.net/MainPage.ashx).

**Reply by MarcinZiabek** [reply]
I understand, no worries. Thank you for the collaboration. Unfortunately, you are the first to experience this problem on Windows. 

If you could create a minimal example of your solution (GIT repository), it could help me investigate and resolve the problem for others.

That being said, I agree that this needs to be improved and fixed. My highest priority is to build QuestPDF-specific skia, simplify the platform and get higher control over the release process. I hope that it will be much better soon.

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. We apologize for the delayed response. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!

**Reply by bigpilot** [reply]
No problem. I've moved to another PDF library.
